pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -158,7 +152,9 @@ func (w *Watcher) invokeRegistrationCallbackAtHandler(ctx context.Context, clien
 	}); err != nil {
 		return fmt.Errorf("failed to send registration status at socket %s, err: %v", socketPath, err)
 	}
-	chanForAckOfNotification <- true
+
+	handler.Register(infoResp.Name, infoResp.Endpoint)
```

Comment:
"Register" after "NotifyRegistrationStatus" might be confusing. May be we can try more intutive names.

Reply:
Do you have any suggestions?

Reply:
The reasoning I had for choosing these names was the following, the consumer of the API is not a plugin but another module of kubelet.

Therefore, it seemed intuitive to name the method as the actions that are going to be taken by the consumer. i.e: Validate the plugin and Register the plugin.

Reply:
`Validate --> InitiateRegistration` it's not what a consumer of the API is expected to do, here Validate makes a lot of sense.

`Register -->CompleteRegistration` if we keep Validate that name doesn't fit anymore :)

Reply:
>  it's not what a consumer of the API is expected to do, here Validate makes a lot of sense.
I expect a consumer api to register plugin in this call and just for handshake completion notification before using registered plugin service.
`Register`  is simply notification of handshake completion. Parameters in the `Register` are also not required and redundant because all that info is already passed in first call. 
I like the idea of using interface but i still think channels suit well for signal passing.

Reply:
> I expect a consumer api to register plugin in this call and just for handshake completion notification before using registered plugin service.

That's the behavior this PR is trying to prevent by making it explicit when the registration should be done.

Here we are not signal passing but transitioning state. Channels are not meant as a mechanism for a state machine, they aren't maintainable (you can't easily add more state) and here introduce a non explicit API.

> Parameters in the Register are also not required and redundant because all that info is already passed in first call

This allows you to choose not to have to keep state.

Reply:
separating validation and registration in two different calls seems to be introducing a race because of the time gap between the api calls.

Reply:
sure, two plugins drops their listening socket with same name at same time. Validation get passed for both and both gets success in `NotifyRegistrationStatus`, which should not happen. One of them should fail at validation.

Reply:
> sure, two plugins drop their listening socket with same name at same time. Validation get passed for both and both gets success in NotifyRegistrationStatus, which should not happen. One of them should fail at validation.

"two plugins drop their listening socket with same name at same time" or "two plugins drop their listening socket with same name at one second interval" is the same issue.

The behavior you choose to take should not be part of the pluginwatcher lifecycle (i.e: Validate, NotifyRegistrationStatus, ...). It would make more sense for you to handle that in the pluginwatcher consumer lifecycle (i.e: Register)

And that's what we are already doing in the deviceplugin with the re-registration mechanism. So I don't think that's really a race or if it is, this PR isn't introducing it as it's a behavior already present.

Reply:
> "two plugins drop their listening socket with same name at same time" or "two plugins drop their listening socket with same name at one second interval" is the same issue.

I think these are different. First case can hit [this condition](https://github.com/kubernetes/kubernetes/pull/58755/files#diff-b54ee8195220ea88e59b6eb423e83dd1R401) but second will not.

> The behavior you choose to take should not be part of the pluginwatcher lifecycle (i.e: Validate, NotifyRegistrationStatus, ...). It would make more sense for you to handle that in the pluginwatcher consumer lifecycle (i.e: Register)

IMO, complete registration should be in scope of plugin watcher and any errors occuring during registration should be notified to the plugin. After success notification to the plugin, there should be no registration related errors.

> So I don't think that's really a race or if it is, this PR isn't introducing it as it's a behavior already present.

After this PR, i am not sure how i will be able to catch and notify plugin about the occurrence of the first case that i referred above.

Reply:
Hello @vikaschoudhary16 !

Your statement very paradoxical as:
- if complete registration is in the scope of the plugin watcher then you don't need to "notify plugin about the occurrence of the first case that you referred above", as it's taken care of by the pluginwatcher.

- If complete registration isn't in the scope then just like the current behavior of the device manager, you don't notify the plugin of it.

Now checking the current implementation that you pointed at I'm a bit surprised by it. This seem like a fix that should have been reviewed a bit more in depth on at least two levels, ....

**On the design level:**
Suppose the two following scenarios:
- Plugin `A` and `A-new` register one after the other, triggering the re-registration code path
   - Plugin `A-new` is stopped and plugin `A` is the plugin that is kept 
- Plugin `A` and `A-new` register one after the other, with a 5 second delay
   - Plugin `A` is stopped and plugin `A-new` is the plugin that is kept 

The behavior is pretty inconsistent for something that clearly seems like the same scenario.

**On the implementation level:**
This whole `if ext != old {` would be completely avoided by creating the endpoint as the first step of the function and moving the device copy under a single lock.

I'm tackling this implementation detail in the related PR.

Finally if complete registration is in the scope of the pluginwatcher then in that case Initialization of the device plugin (e.g: call to `GetDevicePluginOptions` and the first `ListWatch` call) is in the scope and in that case we will need to define more clearly and make changes to the device plugin code.

In any case for complete registration this is clearly not the PR to discuss this, please open an issue to discuss that as this PR does not change the behavior

Reply:
> if complete registration is in the scope of the plugin watcher then you don't need to "notify plugin about the occurrence of the first case that you referred above", as it's taken care of by the pluginwatcher.

By "complete registration in scope" i meant that "From the kubelet side, all communication with plugin until registration is not complete will be handled by pluginwatcher. Consumer will take over once registration is done."

> Plugin A and A-new register one after the other, triggering the re-registration code path

**In current implementation i.e without probe:**
since `addEndpoint` is invoked in a goroutine for each register, even if lock is taken just at the start of `addEndpoint`, which of the two threads will take lock first cannot be determined with guarantee. Therefore i dont think even if reviewed in greater depth, solution that you proposed would have worked.

**Now after plugin watcher i.e probe mode:**
Here also each socket discovered is processed in a separate goroutine and because of that it may happen that goroutine threads may not follow the order in which plugins were started, if the interval is too small.

Reply:
I like the overall idea of explicit interface, but i also think there still design questions need be sorted out.
The handler has `Validate()` and `Register()`

1)  `Register()` can still fail, but i do not see it been handled anywhere, do i read it correct?

2) if  `Validate()` need to read any data that will be written by `Register()`, then it going to be a race condition.
for example, if csi handler trying to detect duplicate name registration in `Validate()`, then `Validate()` need to read `csiDrivers.driversMap`, it is going to a race condition.

the use of `chanForAckOfNotification ` is for handler make sure registration status been send to client successfully, if not, handler can has a follow up,  but what kind of follow up it will has? maybe It is a good time to re-visit it, 

if it is not really needed (csi handler does not implement the channel nor taken any follow up action yet), then we can either remove it or keep it for future ( and we don't have to worry too much about it)

Thanks

Reply:
> By "complete registration in scope" i meant that "From the kubelet side, all communication with plugin until registration is not complete will be handled by pluginwatcher. Consumer will take over once registration is done."

Yes, my point is if "complete registration in scope" of the pluginwatcher then "you" (i.e: the pluginwatcher consumer) should not "notify plugin about the occurrence of the first case that you referred above" because the pluginwatcher is doing that.
Meaning the race described above doesn't happen.


> Therefore i dont think even if reviewed in greater depth, solution that you proposed would have worked.
> because of that it may happen that goroutine threads may not follow the order in which plugins were started, if the interval is too small.

The order in which the plugins were started is not the point. The point is the order in which the devices are registered. The current codepath basically handles these two scenarios in two different ways.
In any case, let's take that argument to your PR.

> Register() can still fail, but i do not see it been handled anywhere, do i read it correct?

Yes at this point, its pointless to handle it in the pluginwatcher (since it can't be bubbled up to the plugin). The consumer might as well log any error it sees and return.